components/patient-data/rest/src/test/java/org/phenotips/data/rest/internal/DefaultPatientsResourceImplTest.java

```diff
@@ -0,0 +1,199 @@
+        doReturn(executionContext).when(execution).getContext();
+        doReturn(mock(XWikiContext.class)).when(executionContext).getProperty("xwikicontext");
+
+        this.repository = this.mocker.getInstance(PatientRepository.class);
+        this.users = this.mocker.getInstance(UserManager.class);
+        this.access = this.mocker.getInstance(AuthorizationManager.class);
+        this.patientsResource = (DefaultPatientsResourceImpl)this.mocker.getComponentUnderTest();
+        this.logger = this.mocker.getMockedLogger();
+        this.queries = this.mocker.getInstance(QueryManager.class);
+        this.uri = new URI("http://uri");
+
+        doReturn(this.uri).when(this.uriInfo).getBaseUri();
+        doReturn(this.uri).when(this.uriInfo).getRequestUri();
+        ReflectionUtils.setFieldValue(this.patientsResource, "uriInfo", this.uriInfo);
+
+        doReturn("00000001").when(this.patient).getId();
+        doReturn(this.currentUser).when(this.users).getCurrentUser();
+        doReturn(null).when(this.currentUser).getProfileDocument();
+    }
+
+    @Test
+    public void addPatientUserDoesNotHaveAccess() throws XWikiRestException {
+        WebApplicationException exception = new WebApplicationException();
+        doReturn(false).when(this.access).hasAccess(Right.EDIT, null, mock(EntityReference.class));
+        try {
+            Response response = this.patientsResource.addPatient("");
+        }
+        catch (WebApplicationException ex){
+            exception = ex;
+        }
+        Assert.assertEquals(Response.Status.UNAUTHORIZED.getStatusCode(), exception.getResponse().getStatus());
+        verify(this.logger).debug("Importing new patient from JSON via REST: {}", "");
+    }
+
+    @Test
+    public void addNullPatient() throws XWikiRestException {
+        doReturn(true).when(this.access).hasAccess(eq(Right.EDIT), any(DocumentReference.class), any(EntityReference.class));
+        Response response = this.patientsResource.addPatient(null);
+        Assert.assertEquals(Response.Status.BAD_REQUEST.getStatusCode(), response.getStatus());
+        verify(this.logger).error(eq("Could not process remote matching request: {}"), anyString(), anyObject());
+    }
+
+    @Test
+    public void addPatientAsJSON() throws XWikiRestException {
+        doReturn(true).when(this.access).hasAccess(eq(Right.EDIT), any(DocumentReference.class), any(EntityReference.class));
+        JSONObject json = new JSONObject();
+        Response response = this.patientsResource.addPatient(json.toString());
+        verify(this.logger).debug("Importing new patient from JSON via REST: {}", json.toString());
+    }
+
+    @Test
+    public void listPatientsNullOrderField() throws XWikiRestException {
+        WebApplicationException exception = new WebApplicationException();
+        try {
+            Patients result = this.patientsResource.listPatients(0, 30, null, "asc");
+        }
+        catch (WebApplicationException ex){
```

Comment:
Yes, @DeanWay is right, there shouldn't be an exception when these are missing. Don't write tests that validate the current implementation, but think about the expected behavior and test that.

Tests are another way of providing a specification. The method name should be a short sentence describing what happens when calling a method within a specific environment/configuration. The mocks setup calls should clearly identify what that configuration is, and the assertions should show what's expected to happen — or NOT happen. The mocks should reflect a sane environment, even if the tested code path doesn't need every part of that environment.